ArduPilot/ardupilot
# Custom Controller Support 2nd Version

This PR has the same aim with the following PR [Custom Controller Support #21123
](https://github.com/ArduPilot/ardupilot/pull/21123), 

This PR has the following feature flag to compile out the changes in original code.
- In-flight run time transition between original and custom controller with RC switch.
- Proper copter rate controller reset to provide bumpless transition from custom controller
- Ground and inflight spool state separation to avoid build-up during arming and takeoff with the custom controller
- bitmask to choose which axis to run the custom controller for

I also added an example empty and PID controller. PID controller has the same architecture as the copter, but gains are scaled by 90%. In the example, the custom controller gets its already input-shaped target from the AC_AttitudeControl library. 

I also find that not running the original controller means that ATC_THR_MIX_MIN, ATC_THR_MIX_MAN, ATC_THR_MIX_MAX have no effect during different stages of flight if we are flying with a custom controller. Because update_throttle_rpy_mix function runs inside rate_controller_run. So, it might be a good idea to run the primary controller regardless or move update_throttle_rpy_mix out of rate_controller_run function.

Pros and cons I find with this approach.

Pros:
- Easily access the attitude controller and all of its functionality
- No modification to the AC_AttitudeController code and small modification to the vehicle code
- Systematic way to store multiple new controllers that might be introduced in the future
- Complete separation between custom controller and AttitudeControl libraries

Cons:
- Increase flash space with each custom controller, 

How to test:

Set CUSTOMCONTROL_ENABLED inside AC_CustomControl.h to 1
Set CSC_CONT_TYPE parameter to 2(working PID controller)
Assign RCx_OPTION to 109. Ex: RC6_OPTION 109
arm and takeoff
Switch RCx to high value
No difference can be observed as expected while switching. A message will pop on the ground control station when switching between controllers.
